src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,337 @@
+        }
+
+        /// <inheritdoc />
+        public async Task InitializeAsync()
+        {
+            await this.provenBlockHeaderRepository.InitializeAsync().ConfigureAwait(false);
+
+            this.TipHashHeight = this.provenBlockHeaderRepository.TipHashHeight;
+
+            this.logger.LogDebug("Initialized ProvenBlockHeader block tip at '{0}'.", this.TipHashHeight);
+
+            this.asyncLoop = this.asyncLoopFactory.Run("ProvenBlockHeaders job", async token =>
+            {
+                // Save pending items.
+                await this.SaveAsync().ConfigureAwait(false);
+
+                this.logger.LogTrace("(-)[IN_ASYNC_LOOP]");
+            },
+            this.nodeLifetime.ApplicationStopping,
+            repeatEvery: TimeSpan.FromMinutes(1),
+            startAfter: TimeSpan.FromMinutes(1));
+        }
+
+        /// <inheritdoc />
+        public async Task<ProvenBlockHeader> GetAsync(int blockHeight)
+        {
+            ProvenBlockHeader header = null;
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddQueryTime(o)))
+            {
+                this.Cache.TryGetValue(blockHeight, out header);
+
+                if (header == null)
+                {
+                    // Check the repository.
+                    header = await this.provenBlockHeaderRepository.GetAsync(blockHeight).ConfigureAwait(false);
+
+                    this.Cache.AddOrUpdate(blockHeight, header, header.HeaderSize);
+                }
+            }
+
+            return header;
+        }
+
+        /// <inheritdoc />
+        public async Task<List<ProvenBlockHeader>> GetAsync(int fromBlockHeight, int toBlockHeight)
+        {
+            Guard.Assert(toBlockHeight >= fromBlockHeight);
+
+            var headersInCache = new List<ProvenBlockHeader>();
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddQueryTime(o)))
+            {
+                int index = fromBlockHeight;
+
+                ProvenBlockHeader header = null;
+
+                var headersNotInCache = new List<int>();
+
+                do
+                {
+                    this.Cache.TryGetValue(index, out header);
+
+                    if (header != null)
+                    {
+                        headersInCache.Add(header);
+                    }
+                    else
+                    {
+                        headersNotInCache.Add(index);
+                    }
+
+                    index++;
+
+                } while (index <= toBlockHeight);
+
+                // Try and get items from the repository if not found in the store cache.
+
+                if (headersNotInCache.Count > 0)
+                {
+                    var repositoryHeaders = new List<ProvenBlockHeader>();
+
+                    // Check the full range first
+                    if ((headersNotInCache.Count - 1) == toBlockHeight - fromBlockHeight)
```

Comment:
won't it be the same if we check headersInCache.Count == 0 ?

Reply:
Actually yes, and this looks more clear.